### PR TITLE
[Merged by Bors] - Add __xstat hook

### LINKED ISSUE #867: stat does not work on linux
### Bug Description

Doing `stat` on a file that exists on the remote pod, does not seem to work.

### Steps to Reproduce

```
#include <stdio.h>
#include <sys/stat.h>
#include <errno.h>
#include <string.h>

int main(int argc, char **argv) {
	struct stat statbuf;
	int err = stat("/app/test.txt", &statbuf);
	if (err != 0) {
		fprintf(stderr, "%s\n", strerror(errno));
		return 1;
	}
	printf("Size: %ld\n", statbuf.st_size);
	printf("Permissions: %o\n", statbuf.st_mode);

	return 0;
}
```

- run the above program with target a remote pod that has "/app/test.txt"

### Backtrace

_No response_

### Relevant Logs

```shell
2022-12-31T02:29:04.075044Z TRACE ThreadId(01) enable_hooks: mirrord_layer: exit enabled_file_ops=true enabled_remote_dns=true
2022-12-31T02:29:04.075286Z TRACE ThreadId(01) start_layer_thread: mirrord_layer: enter config=LayerConfig { accept_invalid_certificates: true, skip_processes: None, target: TargetConfig { path: Some(Pod(PodTarget { pod: "py-serv-deployment-5c57fbdc98-pdbn4", container: None })), namespace: None }, connect_tcp: None, connect_agent_name: Some("mirrord-agent-vnewddxw49-2g6rc"), connect_agent_port: Some(52519), agent: AgentConfig { log_level: "info", namespace: None, image: None, image_pull_policy: "IfNotPresent", ttl: 0, ephemeral: false, communication_timeout: None, startup_timeout: 60, network_interface: None, pause: false }, feature: FeatureConfig { env: EnvConfig { include: None, exclude: None, overrides: None }, fs: FsConfig { mode: Read, include: None, exclude: None, read_write: None, read_only: None, local: None }, network: NetworkConfig { incoming: Mirror, outgoing: OutgoingConfig { tcp: true, udp: true }, dns: true }, capture_error_trace: false }, operator: OperatorConfig { enabled: true, namespace: "mirrord", port: 8080 } }
2022-12-31T02:29:04.075383Z TRACE ThreadId(01) start_layer_thread: mirrord_layer: exit config=LayerConfig { accept_invalid_certificates: true, skip_processes: None, target: TargetConfig { path: Some(Pod(PodTarget { pod: "py-serv-deployment-5c57fbdc98-pdbn4", container: None })), namespace: None }, connect_tcp: None, connect_agent_name: Some("mirrord-agent-vnewddxw49-2g6rc"), connect_agent_port: Some(52519), agent: AgentConfig { log_level: "info", namespace: None, image: None, image_pull_policy: "IfNotPresent", ttl: 0, ephemeral: false, communication_timeout: None, startup_timeout: 60, network_interface: None, pause: false }, feature: FeatureConfig { env: EnvConfig { include: None, exclude: None, overrides: None }, fs: FsConfig { mode: Read, include: None, exclude: None, read_write: None, read_only: None, local: None }, network: NetworkConfig { incoming: Mirror, outgoing: OutgoingConfig { tcp: true, udp: true }, dns: true }, capture_error_trace: false }, operator: OperatorConfig { enabled: true, namespace: "mirrord", port: 8080 } }
No such file or directory
```

### Your operating system and version

Linux xxx 5.15.0-1023-azure #29~20.04.1-Ubuntu SMP Wed Oct 26 19:18:25 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Local process

RUST_BACKTRACE=1 RUST_LOG=mirrord=trace ../mirrord/target/debug/mirrord exec -c -t pod/py-serv-deployment-5c57fbdc98-pdbn4 -- ./a.out

### Local process version

_No response_

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by infiniteregrets**
closes #867

### CODE REVIEW

File: mirrord/layer/src/file/hooks.rs [hunk 1]
```diff
@@ -478,6 +478,25 @@ unsafe extern "C" fn stat_detour(raw_path: *const c_char, out_stat: *mut stat) -
...
 
+/// Hook for libc's stat syscall wrapper.
+#[hook_guard_fn]
+pub(crate) unsafe extern "C" fn __xstat_detour(
+    ver: c_int,
+    raw_path: *const c_char,
+    out_stat: *mut stat,
+) -> c_int {
+    let path = (!raw_path.is_null()).then(|| CStr::from_ptr(raw_path));
+    let (Ok(result) | Err(result)) = xstat(Some(ver), Some(path), None, true)
```
Comment (aviramha) [inline mirrord/layer/src/file/hooks.rs:489]: I think the version check should be done here as it's not shared logic..
also, according to docs it should be `3`
https://refspecs.linuxfoundation.org/LSB_1.1.0/gLSB/baselib-xstat-1.html
I assume you saw version 1 being used?

Comment (infiniteregrets) [reply mirrord/layer/src/file/hooks.rs:489]: yeah when I traced it, I saw version 1 being used

Comment (infiniteregrets) [reply mirrord/layer/src/file/hooks.rs]: verdict on this? I only saw `1` being used in my runs, do we allow both?

disassembly of the function:
```
int __xstat(int __ver,char *__filename,stat64 *__stat_buf)

{
  undefined4 *in_FS_OFFSET;
  
  if ((uint)__ver < 2) {
    syscall();
    return 4;
  }
  *in_FS_OFFSET = 0x16;
  return -1;
}
```

Comment (aviramha) [reply mirrord/layer/src/file/hooks.rs]: sounds okay

**Review by aviramha — COMMENTED** [review]
great job! 
bors r+

**Review by aviramha — APPROVED** [review]

**Pull request closed without merging**